
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>simulate_endToEnd &#8212; TumorSim  documentation</title>
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for simulate_endToEnd</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">SeqIO</span>
<span class="kn">from</span> <span class="nn">mutation_orchestrator</span> <span class="k">import</span> <span class="n">Mutation_Orchestrator</span>
<span class="kn">import</span> <span class="nn">copy</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">copy</span>

<div class="viewcode-block" id="write_fasta"><a class="viewcode-back" href="../index.html#simulate_endToEnd.write_fasta">[docs]</a><span class="k">def</span> <span class="nf">write_fasta</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">output_fasta_file</span><span class="p">):</span>  
    <span class="sd">&quot;&quot;&quot; Writes a genome object to an output fasta file &quot;&quot;&quot;</span>
    <span class="n">output_seqs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">genome</span><span class="p">:</span>
        <span class="n">output_seq</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">SeqRecord</span><span class="p">(</span><span class="n">seq</span><span class="o">=</span><span class="n">genome</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span><span class="o">.</span><span class="n">toseq</span><span class="p">(),</span>
              <span class="nb">id</span><span class="o">=</span><span class="n">chrom</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
        <span class="n">output_seqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">output_seq</span><span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_fasta_file</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">output_handle</span><span class="p">:</span>
        <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">output_seqs</span><span class="p">,</span> <span class="n">output_handle</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="remove_trailing_N_characters"><a class="viewcode-back" href="../index.html#simulate_endToEnd.remove_trailing_N_characters">[docs]</a><span class="k">def</span> <span class="nf">remove_trailing_N_characters</span><span class="p">(</span><span class="n">sequence</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Strings representing the nucleotides typically start and end with </span>
<span class="sd">        repeating sequences of the &#39;N&#39; character. This function strips them </span>
<span class="sd">        from the right and left side of the input sequence. &quot;&quot;&quot;</span> 
    <span class="n">original_len_seq</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
    <span class="k">while</span> <span class="n">sequence</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;N&#39;</span><span class="p">:</span>
        <span class="n">sequence</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">offset</span> <span class="o">=</span> <span class="n">original_len_seq</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
    <span class="k">while</span> <span class="n">sequence</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;N&#39;</span><span class="p">:</span>
        <span class="n">sequence</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">offset</span><span class="p">)</span></div>

<div class="viewcode-block" id="read_fasta_normal"><a class="viewcode-back" href="../index.html#simulate_endToEnd.read_fasta_normal">[docs]</a><span class="k">def</span> <span class="nf">read_fasta_normal</span><span class="p">(</span><span class="n">input_fasta_file</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Read in an input fasta file that represents a normal (non-cancerous) genome </span>
<span class="sd">        It takes some time to load entire 3GB hg38 into memory; possible performance problem &quot;&quot;&quot;</span>
    <span class="n">genome</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">genome_offset</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">seq_record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">input_fasta_file</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">):</span>
        <span class="n">genome</span><span class="p">[</span><span class="n">seq_record</span><span class="o">.</span><span class="n">id</span><span class="p">],</span> <span class="n">genome_offset</span><span class="p">[</span><span class="n">seq_record</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">remove_trailing_N_characters</span><span class="p">(</span>
            <span class="n">seq_record</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">tomutable</span><span class="p">())</span>

    <span class="c1">## remove all &#39;useless&#39; chromosomes, i.e. must match chrX, chrY or &quot;^[a-z]{3}\d{1,2}$&quot;</span>
    <span class="n">genome</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="p">:</span> <span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">genome</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> 
                    <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s1">&#39;^[a-z]</span><span class="si">{3}</span><span class="s1">\d{1,2}$&#39;</span><span class="p">,</span> <span class="n">k</span><span class="p">,</span> <span class="n">re</span><span class="o">.</span><span class="n">IGNORECASE</span><span class="p">)</span> <span class="ow">or</span> <span class="n">k</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;chrX&quot;</span><span class="p">,</span> <span class="s2">&quot;chrY&quot;</span><span class="p">]}</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">genome_offset</span><span class="p">)</span></div>

<div class="viewcode-block" id="subtract_beds"><a class="viewcode-back" href="../index.html#simulate_endToEnd.subtract_beds">[docs]</a><span class="k">def</span> <span class="nf">subtract_beds</span><span class="p">(</span><span class="n">bed1</span><span class="p">,</span> <span class="n">bed2</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Subtract identical intervals shared with bed2 from bed1 &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">bed1</span><span class="p">[</span><span class="o">~</span><span class="p">(</span><span class="n">bed1</span><span class="p">[</span><span class="s1">&#39;uid&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">bed2</span><span class="p">[</span><span class="s1">&#39;uid&#39;</span><span class="p">]))]</span></div>

<span class="k">def</span> <span class="nf">write_bed</span><span class="p">(</span><span class="n">genome_offset</span><span class="p">,</span> <span class="n">dframe</span><span class="p">,</span> <span class="n">path</span><span class="p">):</span>
    <span class="n">corrected_bed</span> <span class="o">=</span> <span class="n">offset_bed</span><span class="p">(</span><span class="n">dframe</span><span class="p">,</span> <span class="n">genome_offset</span><span class="p">)</span>
    <span class="n">corrected_bed</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">offset_bed</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">genome_offset</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">genome_offset</span><span class="p">:</span>
        <span class="n">per_chrom</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">chrom</span><span class="p">]</span>
        <span class="n">df</span><span class="o">.</span><span class="n">ix</span><span class="p">[</span><span class="n">per_chrom</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="n">genome_offset</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span>
        <span class="n">df</span><span class="o">.</span><span class="n">ix</span><span class="p">[</span><span class="n">per_chrom</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="n">genome_offset</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">df</span>

<div class="viewcode-block" id="create_complementary_genome"><a class="viewcode-back" href="../index.html#simulate_endToEnd.create_complementary_genome">[docs]</a><span class="k">def</span> <span class="nf">create_complementary_genome</span><span class="p">(</span><span class="n">genome</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; This function outputs the 3&#39;-5&#39; complement</span>
<span class="sd">        All of the reference FASTAs written in 5&#39;-3&#39; &quot;&quot;&quot;</span>
    <span class="n">new_genome</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">new_genome</span><span class="p">:</span>
        <span class="n">new_genome</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span><span class="o">.</span><span class="n">complement</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">new_genome</span> </div>

<span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
     <span class="c1"># read genome fasta</span>
    <span class="p">(</span><span class="n">mutated_genome</span><span class="p">,</span> <span class="n">genome_offset</span><span class="p">)</span> <span class="o">=</span> <span class="n">read_fasta_normal</span><span class="p">(</span><span class="n">args</span><span class="p">[</span><span class="s1">&#39;input_fasta&#39;</span><span class="p">])</span>

    <span class="n">orchestrator</span> <span class="o">=</span> <span class="n">Mutation_Orchestrator</span><span class="p">()</span>
    <span class="c1"># add germilne SNVs &amp; InDels</span>
    <span class="n">mutated_genome</span> <span class="o">=</span> <span class="n">orchestrator</span><span class="o">.</span><span class="n">snv_fast</span><span class="p">(</span><span class="n">mutated_genome</span><span class="p">,</span> <span class="n">args</span><span class="p">[</span><span class="s1">&#39;number_snvs&#39;</span><span class="p">])</span>
    <span class="n">orchestrator</span><span class="o">.</span><span class="n">generate_indels</span><span class="p">(</span><span class="n">mutated_genome</span><span class="p">,</span> <span class="n">args</span><span class="p">[</span><span class="s1">&#39;number_indels&#39;</span><span class="p">])</span>
    <span class="p">(</span><span class="n">mutated_genome</span><span class="p">,</span> <span class="n">snv_and_indel_bed</span><span class="p">)</span> <span class="o">=</span> <span class="n">orchestrator</span><span class="o">.</span><span class="n">generate_fasta_and_bed</span><span class="p">(</span><span class="n">mutated_genome</span><span class="p">)</span>
    <span class="c1">### write out &quot;normalsim&quot; bedpe and fasta</span>
    <span class="n">write_fasta</span><span class="p">(</span><span class="n">mutated_genome</span><span class="p">,</span> <span class="n">args</span><span class="p">[</span><span class="s1">&#39;output_normal_fasta&#39;</span><span class="p">])</span>
    <span class="n">write_bed</span><span class="p">(</span><span class="n">genome_offset</span><span class="p">,</span> <span class="n">snv_and_indel_bed</span><span class="p">,</span> <span class="n">args</span><span class="p">[</span><span class="s1">&#39;output_normal_bedfile&#39;</span><span class="p">])</span>  

    <span class="c1">## output complement 3&#39;-5&#39; strand normal</span>
    <span class="n">normal_complement</span> <span class="o">=</span> <span class="n">create_complementary_genome</span><span class="p">(</span><span class="n">mutated_genome</span><span class="p">)</span>
    <span class="n">write_fasta</span><span class="p">(</span><span class="n">normal_complement</span><span class="p">,</span> <span class="n">args</span><span class="p">[</span><span class="s1">&#39;output_complement_normal_fasta&#39;</span><span class="p">])</span>
    <span class="k">del</span> <span class="n">normal_complement</span>

    <span class="c1"># add structural varations</span>
    <span class="n">orchestrator</span><span class="o">.</span><span class="n">generate_structural_variations</span><span class="p">(</span><span class="n">mutated_genome</span><span class="p">,</span> <span class="n">args</span><span class="p">[</span><span class="s1">&#39;number_of_tumorSVs&#39;</span><span class="p">])</span>
    <span class="p">(</span><span class="n">mutated_genome</span><span class="p">,</span> <span class="n">tumor_bed</span><span class="p">)</span> <span class="o">=</span> <span class="n">orchestrator</span><span class="o">.</span><span class="n">generate_fasta_and_bed</span><span class="p">(</span><span class="n">mutated_genome</span><span class="p">)</span>
    <span class="n">write_fasta</span><span class="p">(</span><span class="n">mutated_genome</span><span class="p">,</span> <span class="n">args</span><span class="p">[</span><span class="s1">&#39;output_tumor_fasta&#39;</span><span class="p">])</span>
    <span class="n">write_bed</span><span class="p">(</span><span class="n">genome_offset</span><span class="p">,</span> <span class="n">tumor_bed</span><span class="p">,</span> <span class="n">args</span><span class="p">[</span><span class="s1">&#39;output_tumor_bedfile&#39;</span><span class="p">])</span>  <span class="c1">### write out &quot;tumorsim&quot; bedpe</span>

    <span class="c1">## output complement 3&#39;-5&#39; strand tumor</span>
    <span class="n">tumor_complement</span> <span class="o">=</span> <span class="n">create_complementary_genome</span><span class="p">(</span><span class="n">mutated_genome</span><span class="p">)</span>
    <span class="n">write_fasta</span><span class="p">(</span><span class="n">tumor_complement</span><span class="p">,</span> <span class="n">args</span><span class="p">[</span><span class="s1">&#39;output_complement_tumor_fasta&#39;</span><span class="p">])</span>


<span class="k">if</span> <span class="n">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="s1">&#39;Simulate cancer genomic structural variations&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--input_fasta&#39;</span><span class="p">,</span> 
                        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;../data/subsampled_hg38.fa&quot;</span><span class="p">,</span>
                        <span class="n">help</span><span class="o">=</span><span class="s1">&#39;file path for the input (default genome) fasta&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--output_tumor_fasta&#39;</span><span class="p">,</span>
                        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;outputs/tumorsim.fasta&quot;</span><span class="p">,</span>
                        <span class="n">help</span><span class="o">=</span><span class="s1">&#39;file path for the output tumor (cancer genome) fasta&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--output_normal_fasta&#39;</span><span class="p">,</span>
                        <span class="n">default</span> <span class="o">=</span> <span class="s2">&quot;outputs/normalsim.fasta&quot;</span><span class="p">,</span>
                        <span class="n">help</span><span class="o">=</span><span class="s1">&#39;file path for the output normal (SNV-added) fasta&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--output_complement_tumor_fasta&#39;</span><span class="p">,</span>
                        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;outputs/complement_tumorsim.fasta&quot;</span><span class="p">,</span>
                        <span class="n">help</span><span class="o">=</span><span class="s1">&#39;file path for the output complement 3-5 strand tumor (cancer genome) fasta&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--output_complement_normal_fasta&#39;</span><span class="p">,</span>
                        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;outputs/complement_normalsim.fasta&quot;</span><span class="p">,</span>
                        <span class="n">help</span><span class="o">=</span><span class="s1">&#39;file path for the output complement 3-5 strand normal (SNV-added) fasta&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--number_snvs&#39;</span><span class="p">,</span>
                        <span class="n">default</span> <span class="o">=</span> <span class="mi">3000</span><span class="p">,</span>
                        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;number of single nucleotide variants to add to the normal genome&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--number_indels&#39;</span><span class="p">,</span>
                        <span class="n">default</span> <span class="o">=</span> <span class="mi">4150</span><span class="p">,</span>
                        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;number of small insertions and deletions to add to the normal genome&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--number_of_tumorSVs&#39;</span><span class="p">,</span>
                        <span class="n">default</span> <span class="o">=</span> <span class="mi">10000</span><span class="p">,</span>
                        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;number of structural variations to add to the tumor genome&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--output_normal_bedfile&#39;</span><span class="p">,</span>
                        <span class="n">default</span> <span class="o">=</span> <span class="s2">&quot;outputs/normal.bed&quot;</span><span class="p">,</span>
                        <span class="n">help</span><span class="o">=</span><span class="s1">&#39;file path for the output normal (SNV-added) bedfile&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--output_tumor_bedfile&#39;</span><span class="p">,</span>
                        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;outputs/tumorsim.bed&quot;</span><span class="p">,</span>
                        <span class="n">help</span><span class="o">=</span><span class="s1">&#39;file path for the output tumor (cancer genome) bedfile&#39;</span><span class="p">)</span>
    <span class="n">args</span> <span class="o">=</span> <span class="nb">vars</span><span class="p">(</span><span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">())</span>
    <span class="n">main</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">TumorSim</a></h1>








<h3>Navigation</h3>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="index.html">Module code</a><ul>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017, Nick De Veaux, Evan Biederstedt and Dayanne Martins de Castro.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.6.4</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
    </div>

    

    
  </body>
</html>